src/TorchSharp/AutogradFunction.cs

```diff
@@ -0,0 +1,502 @@
+                                throw new NotImplementedException($"Function {Function<T>.Instance.Name} returned a gradient different that is defined at position {i + 1}, but the corresponding forward input was not a Tensor");
+                        }
+
+                        // Convert back to C++ array
+                        return _applyFuncReturnArray.CreateArrayWithSize(output.Select(p => p?.handle ?? IntPtr.Zero).ToArray());
+                    }
+                }
+
+                /// <summary>
+                /// This function is also passed as a parameter to the unmanaged object, so that when the unmanaged object
+                /// is destroyed, then we will remove the hard reference we have of this node so that the GC can collect it. 
+                /// </summary>
+                private void DeleteNode()
+                {
+                    DeleteNode(this);
+                    Dispose();
+                }
+
+                /// <summary>
+                /// When we create the unmanaged object we need to hold a strong reference to it while we apply the whole process
+                /// of populating it with information, but then once we've attached it to the graph we want to still maintain a reference
+                /// to it so that we can call into it, but not a strong reference so that it will be destroyed. Therefore, we store
+                /// both a shared and a weak pointer to the object, and so once we attach it to a graph, we will dispose the shared ptr.
+                /// </summary>
+                public void DisposeSharedPtr()
+                {
+                    THSAutograd_CSharpNode_disposeSharedPtr(handle);
+                    CheckForErrors();
+                    handle.sharedPtr = IntPtr.Zero;
+                }
+
+                public void Dispose()
+                {
+                    Dispose(true);
+                    GC.SuppressFinalize(this);
+                }
+
+                public void Dispose(bool disposing)
+                {
+                    if (disposing) {
+                        if (handle.sharedPtr != IntPtr.Zero) 
+                            DisposeSharedPtr();
+                        if (handle.weakPtr != IntPtr.Zero) {
+                            THSAutograd_CSharpNode_disposeWeakPtr(handle);
+                            CheckForErrors();
+                            handle.weakPtr = IntPtr.Zero;
+                        }
+
+                        _context?.Dispose();
+                        _context = null;
+                        _applyFuncReturnArray?.Dispose();
+                        _applyFuncReturnArray = null;
+                        _outputCache?.ForEach(t => t?.Dispose());
+                        _outputCache = null;
+                    }
+                }
+
+                ~Node()
+                {
+                    Dispose(true);
```

Comment:
Why are you passing 'true' from a finalizer?

Reply:
My bad, will remove and call the regular Dispose().

Reply:
No, I think you're supposed to pass 'false'

Reply:
https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern#finalizable-types

Reply:
Ah, I see. OK, will fix that.